cpp/src/arrow/util/tracing_internal.cc

```diff
@@ -176,8 +176,23 @@ otel::trace::TracerProvider* GetTracerProvider() {
   static nostd::shared_ptr<otel::trace::TracerProvider> provider = InitializeTracing();
   return provider.get();
 }
+
+struct StorageSingleton {
+  StorageSingleton() : storage_(otel::context::GetDefaultStorage()) {
+    otel::context::RuntimeContext::SetRuntimeContextStorage(storage_);
+  }
+  nostd::shared_ptr<otel::context::RuntimeContextStorage> storage_;
+};
+
 }  // namespace
 
+static StorageSingleton storage_singleton;
```

Comment:
Nit: what about moving the definition inside Attach()?

Reply:
I moved the static definition inside the method (although the method names have changed a little)